// @netlify/plugin-emails/src/handler/testHelpers.ts
export interface IEmailIntegrationConfig {
  NETLIFY_EMAILS_PROVIDER: string;
  NETLIFY_EMAILS_PROVIDER_API_KEY: string;
  NETLIFY_EMAILS_SECRET: string;
  NETLIFY_EMAILS_MAILGUN_DOMAIN?: string;
  NETLIFY_EMAILS_MAILGUN_HOST_REGION?: string;
}

export const invalidConfigs: Array<
  Array<string | Partial<IEmailIntegrationConfig> | string[]>
> = [
  [
    "no mailgun host region",
    {
      NETLIFY_EMAILS_PROVIDER: "mailgun",
      NETLIFY_EMAILS_PROVIDER_API_KEY: "testKey",
      NETLIFY_EMAILS_SECRET: "testSecret",
      NETLIFY_EMAILS_MAILGUN_DOMAIN: "testDomain",
    },
    "NETLIFY_EMAILS_MAILGUN_HOST_REGION",
  ],
  [
    "no mailgun domain",
    {
      NETLIFY_EMAILS_PROVIDER: "mailgun",
      NETLIFY_EMAILS_PROVIDER_API_KEY: "testKey",
      NETLIFY_EMAILS_SECRET: "testSecret",
      NETLIFY_EMAILS_MAILGUN_HOST_REGION: "testRegion",
    },
    "NETLIFY_EMAILS_MAILGUN_DOMAIN",
  ],
  [
    "no provider api key",
    {
      NETLIFY_EMAILS_SECRET: "testSecret",
      NETLIFY_EMAILS_PROVIDER: "sengrid",
    },
    "NETLIFY_EMAILS_PROVIDER_API_KEY",
  ],
  [
    "no provider",
    {
      NETLIFY_EMAILS_SECRET: "testSecret",
      NETLIFY_EMAILS_PROVIDER_API_KEY: "testKey",
    },
    "NETLIFY_EMAILS_PROVIDER",
  ],
  [
    "no provider api key with mailgun",
    {
      NETLIFY_EMAILS_PROVIDER: "mailgun",
      NETLIFY_EMAILS_SECRET: "testSecret",
      NETLIFY_EMAILS_MAILGUN_HOST_REGION: "testRegion",
      NETLIFY_EMAILS_MAILGUN_DOMAIN: "testDomain",
    },
    "NETLIFY_EMAILS_PROVIDER_API_KEY",
  ],
  [
    "no provider with mailgun",
    {
      NETLIFY_EMAILS_PROVIDER_API_KEY: "testKey",
      NETLIFY_EMAILS_SECRET: "testSecret",
      NETLIFY_EMAILS_MAILGUN_HOST_REGION: "testRegion",
      NETLIFY_EMAILS_MAILGUN_DOMAIN: "testDomain",
    },
    "NETLIFY_EMAILS_PROVIDER",
  ],
];
